src/lib/y2storage/filesystems/btrfs.rb

```diff
@@ -123,8 +185,108 @@ def self.default_btrfs_subvolume_path
         Yast::ProductFeatures.GetStringFeature(section, feature)
       end
 
+      # Updates the list of subvolumes for the Btrfs filesystem mounted at root
+      #
+      # Subvolumes are shadowed or unshadowed according to current mount points
+      # in the whole system.
+      #
+      # @see #shadow_btrfs_subvolumes
+      # @see #unshadow_btrfs_subvolumes
+      #
+      # @param devicegraph [Devicegraph]
+      def self.refresh_root_subvolumes_shadowing(devicegraph)
+        filesystem = BlkFilesystem.all(devicegraph).detect { |f| f.root? && f.is?(:btrfs) }
+        return if filesystem.nil?
+
+        filesystem.shadow_btrfs_subvolumes(devicegraph)
+        filesystem.unshadow_btrfs_subvolumes(devicegraph)
+      end
+
+      # Removes current shadowed subvolumes
+      # Only subvolumes that "can be shadowed" will be removed.
+      #
+      # @param devicegraph [Devicegraph]
+      def shadow_btrfs_subvolumes(devicegraph)
+        subvolumes = btrfs_subvolumes.select(&:can_be_auto_deleted?)
+        subvolumes.each do |subvolume|
+          next unless subvolume.shadowed?(devicegraph)
+          shadow_btrfs_subvolume(devicegraph, subvolume.path)
+        end
+      end
+
+      # Creates subvolumes that were previously removed because they were shadowed
+      #
+      # @param devicegraph [Devicegraph]
+      def unshadow_btrfs_subvolumes(devicegraph)
+        auto_deleted_subvolumes.each do |spec|
+          mount_point = btrfs_subvolume_mount_point(spec.path)
+          next if BtrfsSubvolume.shadowed?(devicegraph, mount_point)
+          unshadow_btrfs_subvolume(spec.path)
+        end
+      end
+
     protected
 
+      # Removes a subvolume
+      # The subvolume is cached into {auto_deleted_subvolumes} list
+      #
+      # @param devicegraph [Devicegraph]
+      # @param path [String] subvolume path
+      def shadow_btrfs_subvolume(devicegraph, path)
+        subvolume = find_btrfs_subvolume_by_path(path)
+        return false if subvolume.nil?
+
+        add_auto_deleted_subvolume(subvolume.path, subvolume.nocow?)
+        delete_btrfs_subvolume(devicegraph, subvolume.path)
+        true
+      end
+
+      # Creates a previously shadowed subvolume
+      # The subvolume is removed from {auto_deleted_subvolumes} list
+      #
+      # @param path [String] subvolume path
+      def unshadow_btrfs_subvolume(path)
+        spec = auto_deleted_subvolumes.detect { |s| s.path == path }
+        return false if spec.nil?
+
+        subvolume = create_btrfs_subvolume(spec.path, !spec.copy_on_write)
+        subvolume.can_be_auto_deleted = true
```

Comment:
Another example of the fact that generating a subvolume based on an specification should be a responsibility of `SubvolSpecification` (after killing `Planned::BtrfsSubvolume`).